### PR TITLE
feat(common): async streaming read RPC `*Auth` decorator

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #8937 

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8942)

### CODE REVIEW

File: google/cloud/google_cloud_cpp_rest_internal.cmake [hunk 1]
```diff
@@ -21,6 +21,7 @@ find_package(OpenSSL REQUIRED)
 # the library
 add_library(
     google_cloud_cpp_rest_internal # cmake-format: sort
+    internal/async_streaming_read_rpc_auth.h
```
Comment (dbolduc) [inline google/cloud/google_cloud_cpp_rest_internal.cmake:24]: seems wrong

File: google/cloud/google_cloud_cpp_rest_internal.bzl [hunk 2]
```diff
@@ -17,6 +17,7 @@
 """Automatically generated source lists for google_cloud_cpp_rest_internal - DO NOT EDIT."""
 
 google_cloud_cpp_rest_internal_hdrs = [
+    "internal/async_streaming_read_rpc_auth.h",
```
Comment (dbolduc) [inline google/cloud/google_cloud_cpp_rest_internal.bzl:20]: seems wrong

File: google/cloud/internal/async_streaming_read_rpc_auth.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    std::unique_ptr<grpc::ClientContext> ReleaseInitialContext() {
+      std::lock_guard<std::mutex> g{mu};
+      return std::move(initial_context);
+    }
+
+    future<bool> OnStart(
+        StatusOr<std::unique_ptr<grpc::ClientContext>> context) {
+      std::lock_guard<std::mutex> g{mu};
+      if (cancelled) return make_ready_future(false);
```
Comment (dbolduc) [inline google/cloud/internal/async_streaming_read_rpc_auth.h:82]: `Cancel()` is untested. I think we want a test for the case where `Cancel()` is called before the `AsyncConfigureContext()` future is ready.

File: google/cloud/internal/async_streaming_read_rpc_auth.h [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      if (context) {
+        stream = factory(*std::move(context));
+      } else {
+        stream = absl::make_unique<AsyncStreamingReadRpcError<Response>>(
+            std::move(context).status());
+      }
```
Comment (dbolduc) [inline google/cloud/internal/async_streaming_read_rpc_auth.h:88]: I think we also want a test case for when the auth strategy returns a bad status

**Review by dbolduc — CHANGES_REQUESTED** [review]

File: google/cloud/google_cloud_cpp_rest_internal.cmake (hunk 1 above)
Comment (dbolduc) [reply google/cloud/google_cloud_cpp_rest_internal.cmake:24]: Aside: it is surprising to me that our builds do not catch this...

File: google/cloud/google_cloud_cpp_rest_internal.bzl (hunk 2 above)
Comment (coryan) [reply google/cloud/google_cloud_cpp_rest_internal.bzl]: Fixed

File: google/cloud/internal/async_streaming_read_rpc_auth.h (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_auth.h:82]: Done.

**Review by dbolduc — APPROVED** [review]

File: google/cloud/google_cloud_cpp_rest_internal.cmake (hunk 1 above)
Comment (dbolduc) [reply google/cloud/google_cloud_cpp_rest_internal.cmake]: Thanks for explaining

**Pull request merged**